docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+  reasonable to set `usize` and `isize` length to 252 bit,
+- 256 bit is the next power-of-2 after 252. Having `usize` and `isize` 256 bit long leaves 4 extra
+  bits that may be used to keep some metadata.
+
+Ultimately the size of `usize` and `isize` has been decided to be 64 bits, which is neither of the
+above possibilities. This length is a consequence of using the `aarch64-unknown-none-softfloat`
+target triple. The choice of the triple determines the length of the pointer which in turn
+determines the length of `usize`. This target triple is a temporary choice before a custom target
+triple is proposed. It has been chosen for its soft float support and no host operating system. The
+pointer length is just one of its parameters we accept on this stage of the project.
+
+Summing up, we expect to see in the IR integers of the following lengths: 1, 8, 16, 32, 64 and 128
+bits.
+
+#### Pointers
+
+LLVM IR has only one generic pointer type - `ptr`, which works as a rough equivalent of the `void *`
+in C. Type-specific pointers (e.g. equivalent of C's `int`) existed, but are now deprecated.
+Therefore, we expect to see in the input IR only the generic `ptr` pointer.
+
+Not only Rust and the Cairo language support a generic pointer type, their support for pointer does
+not follow the mechanism known from C, which would smoothly translate into LLVM IR semantics. Both
+languages operates on strongly typed smart pointers, that handle, under the hood, their own memory.
+
+Based on these observations, ALU operations must be strongly typed, by the limitation of the
+language it will be implemented in. When the input IR will be parsed, it generic pointer types must
+be inferred from the context, to allow matching with the proper implementation.
+
+#### Vectors
+
+Neither the Cairo VM, Cairo language nor no-std Rust have support for vector operations.
+
+LLVM IR has vectors as first class citizens. However,
+_[vector types are used where multiple primitive data are operated in parallel using a single instruction (SIMD)](https://llvm.org/docs/LangRef.html#t-vector)_.
+If Cairo target definition supplied to `rustc` will not suggest the existence of vector extension on
+the target platform, we do not expect any vector intrinsics to appear in the IR. Therefore, vector
+support is not planned in the initial phase of the project.
+
+#### Type conversion
+
+Cairo does not have Rust's `as` keyword, so it's not possible to do e.g. `let a = b as u32` given
+`b` is a `u64`.
+
+An equivalent operation in Cairo is `let a: u32: b.try_into().unwrap();`. This approach has two
+disadvantages:
+
+- it will panic if the value of `b` is larger than `0xFFFFFFFF`,
+- there is no automatic wraparound as in the case of `as`.
+
+Should type conversion be necessary in the implementation of the operations, it will need to handle
+the type conversion with `try_into()` and manual recover from errors:
```

Comment:
```suggestion
the type conversion with `try_into()` and manually recover from errors:
```

I'm not sure I understand this point. Our conversion operations are extension and truncation which just mess around with bits.